main.cpp

```diff
@@ -13,145 +11,295 @@
+
+  std::vector< std::vector<double> > return_data() const;
+  std::vector<double> return_times() const;
+  Model::Model return_model() const;
+  std::vector<double> return_initial_condition() const;
+  Histograms::Parameters return_histogram_parameters() const;
+  bool within_bounds() const;
+  Sample perturb() const;
+  double perturb_ratio() const;
+
+  Sample operator = (const Sample &sample);
+  operator std::valarray<double> () const;
+
+  friend std::ostream & operator<< (std::ostream &out, const Sample &sample);
+
+private:
+  ConstantData const_parameters;
+};
+
 
 
-// A function that given a SampleType object (i.e., a set of model
-// parameters) computes the corresponding likelihood
-double log_likelihood (const SampleType &prm)
+Sample::Sample(double kb, double k1, double k2, double k3, double k4, unsigned int cutoff)
+  : kb(kb), k1(k1), k2(k2), k3(k3), k4(k4), cutoff(cutoff)
+{}
+
+
+
+Sample::Sample()
+  : Sample(std::numeric_limits<double>::signaling_NaN(),
+           std::numeric_limits<double>::signaling_NaN(),
+           std::numeric_limits<double>::signaling_NaN(),
+           std::numeric_limits<double>::signaling_NaN(),
+           std::numeric_limits<double>::signaling_NaN(),
+           static_cast<unsigned int>(-1))
+{}
+
+
+
+std::vector<std::vector<double>> Sample::return_data() const
 {
-  // data used for likelihood
-  const Data::PomData raw_data;
-  const std::vector<std::valarray<double>> data_atoms = { std::pow(raw_data.tem_diam_time1/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time2/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time3/0.3000805, 3),
-                                                          std::pow(raw_data.tem_diam_time4/0.3000805, 3)};
+  return const_parameters.data_size;
+}
 
-  // create model
-  Models::TwoStepAlternative model;
 
-  // set up initial conditions and times for the ODE solution
-  std::valarray<double> initialCondition(0.0,prm.n_variables);
-  initialCondition[0] = 0.0012;
-  const std::vector<double> times = {0,
-                                     raw_data.tem_time1,
-                                     raw_data.tem_time2,
-                                     raw_data.tem_time3,
-                                     raw_data.tem_time4};
 
+std::vector<double> Sample::return_times() const
+{
+  return const_parameters.times;
+}
 
-  // define parameters for the histogram
-  const Histograms::Parameters histogramParameters(25, model.getSmallestParticleSize(prm), model.getLargestParticleSize(prm));
 
-  // calculate log likelihood and return it
-  return Statistics::log_likelihood(data_atoms, times, model, prm, initialCondition, histogramParameters);
 
+// Here we implement a Four-Step Model
+Model::Model Sample::return_model() const
+{
+  Model::TermolecularNucleation nucleation(const_parameters.A_index, const_parameters.As_index,
```

Comment:
Yes, this won't work. You're creating a local variable here, which will die at the end of the function's lifetime.

Reply:
With the patch in #55, you'll need to write this as follows:
```suggestion
  std::shared_ptr<Model::TermolecularNucleation> nucleation
  = std::make_shared<Model::TermolecularNucleation>(const_parameters.A_index, const_parameters.As_index,
```
or even simpler:
```suggestion
  auto nucleation
  = std::make_shared<Model::TermolecularNucleation>(const_parameters.A_index, const_parameters.As_index,
```